packages/apollo-server-codegen/src/IR/Resolvers.ts

```diff
@@ -0,0 +1,168 @@
+import {
+  FieldDefinitionNode,
+  FieldNode,
+  InputValueDefinitionNode,
+  OperationDefinitionNode,
+  parse,
+  StringValueNode
+} from "graphql";
+import { Translatable, Translator } from "../Translators";
+import { Description } from "./Descriptions";
+import { TypelessObjectDefinition } from "./Objects";
+import {
+  CompoundType,
+  findRootType,
+  makeType,
+  TypeDefinition,
+  NonNullType
+} from "./Types";
+
+const parseSelections = (source: string) =>
+  (parse(`query { ${source} }`).definitions[0] as OperationDefinitionNode)
+    .selectionSet.selections as (readonly FieldNode[]);
+
+export class ArgumentDefinition implements Translatable {
+  public description: Description;
+  public name: string;
+  public type: TypeDefinition;
+
+  constructor(argumentDefinition: InputValueDefinitionNode) {
+    this.description = new Description(argumentDefinition);
+    this.name = argumentDefinition.name.value;
+    this.type = makeType(argumentDefinition.type);
+    if (argumentDefinition.defaultValue) this.type = new NonNullType(this.type);
+  }
+
+  public translate(translator: Translator) {
+    return translator.translateArgumentDefinition(this);
+  }
+}
+
+/**
+ * Shell class to provide info on @provides directives until we can bootstrap with global type definitions.
+ * Bootstrap into a `ResolverDefinition` by calling `applyGlobalTypeKnowledge` once all types are known.
+ */
+export class TypelessResolverDefinition {
+  constructor(
+    public fieldDefinition: FieldDefinitionNode,
+    public parent: TypelessObjectDefinition,
+    public isRootType: boolean
+  ) {}
+
+  public getName() {
+    return this.fieldDefinition.name.value;
+  }
+
+  public getProvides(types: TypelessObjectDefinition[], errors: string[]) {
+    const provideDirectives = (this.fieldDefinition.directives || []).filter(
+      directive =>
+        directive.name.value === "provides" &&
+        directive.arguments &&
+        directive.arguments.length
+    );
+
+    const providedSelections = provideDirectives
```

Comment:
This seems like we should tighten up the type definitions to avoid the casting below this line

Reply:
Got this for free when moving over to the shared util implementation